File: tests/rpc_test.py
import grpc
import beagle_pb2 as pb2

def run():
    # Create a connection to the server
    channel = grpc.insecure_channel('localhost:50051')
    stub = pb2.BeagleboneStub(channel)

    # Create a Jetson_to_beagle message
    jetson_msg = pb2.Jetson_to_beagle()
    jetson_msg.keyboard_cmd = 123
    jetson_msg.xf.extend([1.0, 2.0, 3.0])

    # Create a request with the Jetson_to_beagle message
    request = pb2.BeagleRequest()
    request.jetson_to_beagle.CopyFrom(jetson_msg)

    # Send the request to the server
    response = stub.Beagle(request)

    # Print the Logging_Packet message received from the server
    for log in response.logs:
        print("Received log with time", log.time)
        # Process other fields here as needed

if __name__ == '__main__':
    run()

File: tests/rpc_test.cpp
#include <iostream>
#include <memory>
#include <string>

#include <grpcpp/grpcpp.h>
#include "<filename>.pb.h"
#include "<filename>.grpc.pb.h"

class BeagleboneServiceImpl final : public Beaglebone::Service {
public:
  grpc::Status Beagle(grpc::ServerContext* context, const BeagleRequest* request,
                      Logging_Packet* response) override {
    // Process the Jetson_to_beagle message received from the client
    const Jetson_to_beagle& jetson_msg = request->jetson_to_beagle();
    std::cout << "Received keyboard_cmd: " << jetson_msg.keyboard_cmd() << std::endl;
    for (float xf : jetson_msg.xf()) {
      std::cout << "Received xf: " << xf << std::endl;
    }
    
    // Create and fill in a Logging_Packet message to send back to the client
    Logging* log = response->add_logs();
    log->set_time(123.45);
    // Set other fields as needed

    return grpc::Status::OK;
  }
};

void RunServer() {
  std::string server_address("0.0.0.0:50051");
  BeagleboneServiceImpl service;

  grpc::ServerBuilder builder;
  builder.AddListeningPort(server_address, grpc::InsecureServerCredentials());
  builder.RegisterService(&service);

  std::unique_ptr<grpc::Server> server(builder.BuildAndStart());
  std::cout << "Server listening on " << server_address << std::endl;
  server->Wait();
}

int main(int argc, char** argv) {
  RunServer();
  return 0;
}

